### PR TITLE
Remove unnecessary suppressions in APICompat files

### PR DESCRIPTION
**Pull request by ViktorHofer**
Related https://github.com/dotnet/sdk/pull/32964

- Remove unnecessary suppressions in APICompat files. This is in preparation for https://github.com/dotnet/sdk/pull/32964 which will validate the existing suppressions going forward.
- Set the newly introduced APICompat properties.

### CODE REVIEW

File: eng/resolveContract.targets [hunk 1]
```diff
@@ -62,6 +63,22 @@
...
+  </PropertyGroup>
+
+  <!-- Disable suppression validation for packable projects that have a corresponding reference source project.
+       That is necessary as APICompat is invoked twice, once for the ref <-> src comparision and then again
+       for the package validation (which doesn't include reference assemblies). As both operations don't have
+       all the inputs available, some suppressions might only apply to one or the other and hence unnecessary
+       suppressions can't be determined. -->
+  <PropertyGroup Condition="'$(IsPackable)' == 'true' and '$(IsRuntimeAndReferenceAssembly)' != 'true'">
+    <ApiCompatPreserveUnnecessarySuppressions>true</ApiCompatPreserveUnnecessarySuppressions>
+    <ApiCompatPermitUnnecessarySuppressions>true</ApiCompatPermitUnnecessarySuppressions>
```
Comment (carlossanlop) [inline eng/resolveContract.targets:79]: These two properties sound very similar. Why two?

File: eng/resolveContract.targets [hunk 2]
```diff
@@ -62,6 +63,22 @@
     <_ApiCompatLibReplacementString Condition="'$(MSBuildProjectName)' == 'System.Private.CoreLib'">lib/$(NetCoreAppCurrent)/$2</_ApiCompatLibReplacementString>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <IsRuntimeAndReferenceAssembly Condition="'$(IsRuntimeAndReferenceAssembly)' == '' and
+                                              '$(HasMatchingContract)' != 'true' and
```
Comment (carlossanlop) [inline eng/resolveContract.targets:68]: For my own education: What is a "matching contract"?

File: eng/resolveContract.targets [hunk 3]
```diff
@@ -62,6 +63,22 @@
...
+                                              '$(IsPrivateAssembly)' != 'true'">true</IsRuntimeAndReferenceAssembly>
```
Comment (carlossanlop) [inline eng/resolveContract.targets:69]: Can you give me an example of a private assembly?

File: src/libraries/Directory.Build.targets [hunk 4]
```diff
@@ -82,7 +76,7 @@
   <!-- Libraries-specific binplacing properties -->
   <PropertyGroup>
     <BinPlaceRef Condition="'$(BinPlaceRef)' == '' and ('$(IsReferenceAssemblyProject)' == 'true' or '$(IsRuntimeAndReferenceAssembly)' == 'true')">true</BinPlaceRef>
-    <BinPlaceRuntime Condition="'$(BinPlaceRuntime)' == '' and ('$(IsSourceProject)' == 'true' or '$(IsRuntimeAndReferenceAssembly)' == 'true')">true</BinPlaceRuntime>
+    <BinPlaceRuntime Condition="'$(BinPlaceRuntime)' == '' and '$(IsSourceProject)' == 'true'">true</BinPlaceRuntime>
```
Comment (carlossanlop) [inline src/libraries/Directory.Build.targets:79]: I see that you moved `IsRuntimeAndReferenceAssembly to `resolveContract.targets` and changed it a bit. Why did you move it there?

Can you please explain why you removed the condition from `IsNETCoreAppSrc` and `BinPlaceRuntime`?

File: src/libraries/apicompat/ApiCompatBaseline.netstandard2.1.xml [hunk 5]
```diff
@@ -853,12 +853,6 @@
     <Left>netstandard2.1/netstandard.dll</Left>
     <Right>net8.0/netstandard.dll</Right>
   </Suppression>
-  <Suppression>
-    <DiagnosticId>CP0015</DiagnosticId>
-    <Target>T:System.Runtime.CompilerServices.AsyncMethodBuilderAttribute:[T:System.AttributeUsageAttribute]</Target>
-    <Left>netstandard2.1/netstandard.dll</Left>
-    <Right>net8.0/netstandard.dll</Right>
-  </Suppression>
```
Comment (carlossanlop) [inline src/libraries/apicompat/ApiCompatBaseline.netstandard2.1.xml:861]: Can you please explain why all these suppressions got removed if both `ApiCompatPreserveUnnecessarySuppressions` and `ApiCompatPermitUnnecessarySuppressions` are set to `true` in `resolveContract.targets`?

**Review by carlossanlop — APPROVED** [review]

File: src/libraries/apicompat/ApiCompatBaseline.netstandard2.1.xml (hunk 5 above)
Comment (ViktorHofer) [reply src/libraries/apicompat/ApiCompatBaseline.netstandard2.1.xml:861]: These two properties are conditionally set to true for `IsSourceProject` projects (which apicompat.proj isn't): https://github.com/dotnet/runtime/blob/a9f3262175253dd69eea72007347cd0cda9e6af0/Directory.Build.props#L403

In general, we want to disable these new features for packable projects that also have a matching reference source project. I tried to capture that in the added comment in resolveContract.targets.

File: src/libraries/Directory.Build.targets (hunk 4 above)
Comment (ViktorHofer) [reply src/libraries/Directory.Build.targets:79]: Because `IsruntimeAndReferenceAssembly` is used in a condition in resolveContract.targets and resolveContract.targets is imported before src/libraries/Directory.Build.targets.

File: eng/resolveContract.targets (hunk 3 above)
Comment (ViktorHofer) [reply eng/resolveContract.targets:69]: System.Private.Uri

File: eng/resolveContract.targets (hunk 2 above)
Comment (ViktorHofer) [reply eng/resolveContract.targets:68]: A matching contract is a reference source project assembly in the context of a source project, i.e. ref/net8.0/System.Text.Json.dll is the matching contract assembly for lib/net8.0/System.Text.Json.dll.

File: eng/resolveContract.targets (hunk 1 above)
Comment (ViktorHofer) [reply eng/resolveContract.targets:79]: APICompat is used in dotnet/runtime but also in various other repositories. these two new settings are on by default but need to be disabled at least in runtime for a few projects (mentioned that in the code comment in resolveContract.targets). The feature flags are distinct as the latter allows unnecessary suppressions to exist in the suppression files and the former trims out unnecessary suppressions when re-generating a suppression file via the `/p:GenerateAPICompatSuppressionFile=true` switch.
